pkg/executor/sortexec/topn_worker.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sortexec
+
+import (
+	"container/heap"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/pingcap/failpoint"
+	"github.com/pingcap/tidb/pkg/util/chunk"
+	"github.com/pingcap/tidb/pkg/util/memory"
+)
+
+// topNWorker is used only when topn spill is triggered
+type topNWorker struct {
+	workerIDForTest int
+
+	chunkChannel           <-chan *chunk.Chunk
+	fetcherAndWorkerSyncer *sync.WaitGroup
+	errOutputChan          chan<- rowWithError
+	finishChan             <-chan struct{}
+
+	topn       *TopNExec
+	chkHeap    *topNChunkHeap
+	memTracker *memory.Tracker
+}
+
+func newTopNWorker(
+	idForTest int,
+	chunkChannel <-chan *chunk.Chunk,
+	fetcherAndWorkerSyncer *sync.WaitGroup,
+	errOutputChan chan<- rowWithError,
+	finishChan <-chan struct{},
+	topn *TopNExec,
+	chkHeap *topNChunkHeap,
+	memTracker *memory.Tracker) *topNWorker {
+	return &topNWorker{
+		workerIDForTest:        idForTest,
+		chunkChannel:           chunkChannel,
+		fetcherAndWorkerSyncer: fetcherAndWorkerSyncer,
+		errOutputChan:          errOutputChan,
+		finishChan:             finishChan,
+		chkHeap:                chkHeap,
+		topn:                   topn,
+		memTracker:             memTracker,
+	}
+}
+
+func (t *topNWorker) fetchChunksAndProcess() {
+	// Offset of heap in worker should be 0, as we need to spill all data
+	t.chkHeap.init(t.topn, t.memTracker, t.topn.Limit.Offset+t.topn.Limit.Count, 0, t.topn.greaterRow, t.topn.RetFieldTypes())
+	for t.fetchChunksAndProcessImpl() {
+	}
+}
+
+func (t *topNWorker) fetchChunksAndProcessImpl() bool {
+	select {
+	case <-t.finishChan:
+		return false
+	case chk, ok := <-t.chunkChannel:
+		if !ok {
+			return false
+		}
+		defer func() {
+			t.fetcherAndWorkerSyncer.Done()
+		}()
```

Comment:
why calling `t.fetcherAndWorkerSyncer.Done()` here, isn't `` would be call in `(t *topNWorker) run()`?

Reply:
> why calling `t.fetcherAndWorkerSyncer.Done()` here, seems `t.fetcherAndWorkerSyncer.Done()` will be call in `(t *topNWorker) run()`.

Normally, `t.fetcherAndWorkerSyncer.Done()` will not be called in `(t *topNWorker) run()` as there will be no chunks in `t.chunkChannel`. Only when worker panics or exits in advance will there be some chunks left over in the `t.chunkChannel`, and `t.fetcherAndWorkerSyncer.Done()` will be called in `(t *topNWorker) run()` in this circumstance. 

why `calling t.fetcherAndWorkerSyncer.Done()` here: workers could work only when there are chunks in `t.chunkChannel`, and they will stop working when there is no chunk in `t.chunkChannel`, so fetcher could control the work of workers by `t.chunkChannel`. Fetcher push a chunk into `t.chunkChannel`, worker will process the chunk. Fetcher keeps `t.chunkChannel` empty, workers will stop working. So that fetcher could control workers. When spill happens, fetcher needs to ensure that there are no running workers, so it needs a mechanism to know if all workers have processed all chunks in `t.chunkChannel` and stop working. `t.fetcherAndWorkerSyncer` could take on this task.